packages/auth/src/authClients/firebase.ts

```diff
@@ -50,82 +53,88 @@ export const firebase = (client: Firebase): AuthClient => {
     }
     return provider
   }
-  // Firebase auth functions return a goog.Promise which as of 2021-05-12 does
-  // not appear to work with try {await} catch blocks as exceptions are not caught.
-  // This client returns a new standard Promise so that the exceptions can be
-  // caught and no changes are required in common auth code located in AuthProvider.tsx
-  const repackagePromise = (
-    fireBasePromise: Promise<any | string>
-  ): Promise<any | string> => {
-    return new Promise((resolve, reject) => {
-      fireBasePromise
-        .then((result) => resolve(result))
-        .catch((error) => reject(error))
-    })
+
+  const loginWithEmailLink = ({ email, emailLink }: Options) => {
+    if (
+      email !== undefined &&
+      emailLink !== undefined &&
+      firebaseAuth.isSignInWithEmailLink(auth, emailLink)
+    ) {
+      return firebaseAuth.signInWithEmailLink(auth, email, emailLink)
+    }
+    return undefined
   }
 
   return {
     type: 'firebase',
-    client,
-    restoreAuthState: () => repackagePromise(client.auth().getRedirectResult()),
-    login: (
+    client: auth,
+    restoreAuthState: () => {
+      // return a promise that we be await'd on for first page load until firebase
+      // auth has loaded, indicated by the first firing of onAuthStateChange)
+      return new Promise((resolve, reject) => {
+        const unsubscribe = auth.onAuthStateChanged((user: FirebaseUser) => {
+          unsubscribe()
+          resolve(user)
+        }, reject)
+      })
+    },
+    login: async (
       options: oAuthProvider | Options = { providerId: 'google.com' }
     ) => {
       // If argument provided is a string, it should be the oAuth Provider
       // Cast the provider string into the options object
       if (typeof options === 'string') {
         options = { providerId: options }
       }
+
       if (hasPasswordCreds(options)) {
-        return repackagePromise(
-          client
-            .auth()
-            .signInWithEmailAndPassword(
-              options.email as string,
-              options.password as string
-            )
+        return firebaseAuth.signInWithEmailAndPassword(
+          auth,
+          options.email as string,
+          options.password as string
         )
       }
 
+      if (options.providerId === 'emailLink') {
+        return loginWithEmailLink(options)
+      }
+
       const provider = getProvider(options.providerId || 'google.com')
       const providerWithOptions = applyProviderOptions(provider, options)
-      return repackagePromise(
-        client.auth().signInWithPopup(providerWithOptions)
-      )
+
+      return firebaseAuth.signInWithPopup(auth, providerWithOptions)
     },
-    logout: () => repackagePromise(client.auth().signOut()),
+    logout: async () => auth.signOut(),
```

Comment:
Maybe we can just forward the call log
```suggestion
    logout: auth.signOut,
```

Reply:
I tried this originally and just tried it again but doing that causes an error that I think as to do with the value of "this"
```
index-8fbd02b6.js:2335 Uncaught (in promise) TypeError: this._updateCurrentUser is not a function
    at Object.signOut [as logout] (index-8fbd02b6.js:2335)
    at AuthProvider.logOut (AuthProvider.js:147)
```
wrapping it in a dummy function alleviates this problem.